packages/control-state-mixin/control-state-mixin.ts

```diff
@@ -110,26 +65,36 @@ export const ControlStateMixin = <T extends LitBase>(base: T): LitControlState &
           }, 0);
         }
       });
-
-      if (this.autofocus && !this.disabled) {
-        window.requestAnimationFrame(() => {
-          this._focus();
-          this._setFocused(true);
-          this.setAttribute('focus-ring', '');
-        });
-      }
     }
 
     protected update(props: PropertyValues) {
       if (props.has('disabled')) {
-        this._disabledChanged(this.disabled, props.get('disabled'));
+        if (this.disabled) {
+          this._previousTabIndex = this.tabIndex;
+          this.tabIndex = -1;
+        } else if (props.get('disabled')) {
+          if (this._previousTabIndex !== undefined) {
+            this.tabIndex = this._previousTabIndex;
+          }
+        }
       }
 
       if (props.has('tabIndex')) {
-        // save value of tabindex, as it can be overridden to
-        // undefined in case if the element is disabled
-        this._tabIndexValue = this.tabIndex;
-        this._tabIndexChanged(this.tabIndex);
+        const { tabIndex } = this;
+        // When element is disabled, propagate `tabIndex` to focusElement
+        // and then reset it on the host element to disallow interaction.
+        if (this.disabled && tabIndex != null) {
+          if (tabIndex !== -1) {
+            this._previousTabIndex = tabIndex;
+          }
+          // We need this because of resetting tabIndex to null below,
+          // so in `updated()` we have a reference to the old value.
+          props.set('tabIndex', tabIndex);
+
+          // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+          // @ts-ignore
+          this.tabIndex = null;
```

Comment:
Feels counter-intuitive to need these 2 comment lines. Should we just globally disable `@typescript-eslint/ban-ts-ignore` rule if we sometimes need `@ts-ignore`?

Reply:
Thanks, that makes sense. Done.